docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md

```diff
@@ -0,0 +1,56 @@
+- Code-First - Team starts writing the server side api interface code e.g. controllers, DTOs etc. and later generates and Open API document from it.
+
+## Design-First Approach
+
+A Design-First approach means that APIs are treated as "first-class citizens" and everything about a project revolves around the idea that at the end these APIs will be consumed by clients. So based on the business requirements api development team first start describing API designs as an Open API document and collaborate with the stakeholders to gather feedback.
+
+This approach is quite useful if a project is about developing externally exposed set of APIs which will be consumed by partners. As here we first agree upon an API contract (Open API document) that's why expectations on api producer and consumer side remains strictly clear and both side teams can work in parallel as per the pre-agreed API design.
+
+Key Benefits of this approach:
+
+- Early API design feedback.
+- Expectations remains clear on consumer & producer side as both agreed upon an API contract.
+- Development teams can work in parallel.
+- Testing team can use API contracts to do an early test even before a business logic is in place. By looking at different models, paths, attributes and other aspects of the API testing can provide their input which can be very valuable.
+- During an agile development cycle API definitions are not impacted by incremental dev changes.
+- API design is not influenced by actual implementation limitations & code structure.
+- Server side boilerplate code e.g. controllers, DTOs etc. can be auto generated from API contracts.
+- May improve collaboration between API producer & consumer teams.
+
+Planning a Design-First Development:
+
+1. Identify use cases & key services which API should offer.
+2. Identify key stakeholders of API and try to include them during API design phase to get continuous feedback.
+3. Write API contract definitions.
+4. Maintain consistent style for API status codes, versioning, error responses etc.
+5. Encourage peer reviews via pull requests.
+6. Generate server side boilerplate code & client SDKs from API contract definitions.
+
+Important Points to consider:
+
+- If API requirements can changes often during initial development phase than Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
+- It might be worthwhile to first try out your platform specific code generator because it is possible that a particular platform specific code generator might not be able to generate a flexible & maintainable implementation of actual code. For instance If your web framework requires annotations to be present on your controller classes (e.g. for API versioning or authentication), make sure that the code generation tool you use fully supports them.
```

Comment:
> _...might not be able to generate a flexible & maintainable implementation..._

This point is very subjective. Can we be more concrete or objective about what we are guiding people to consider? For example, should we instead advise the audience to consider how much additional work will be required for the generated code to meet done-done criteria?

Reply:
Thanks for the review. 
Updated the section as per your suggestion to bring more clarity. Pleas take a look again. Thanks.